File: hcs_toolbox_py/trans_con.py
"""
Orthogonal transformation to block controllability staircase form.

Original MATLAB: trans_con.m
"""

import numpy as np
from scipy.linalg import null_space


def trans_con(A, B, tol=1e-6):
    """
    Compute an orthogonal matrix T that transforms a controllable pair {A, B}
    to the canonical block controllability staircase form:

        T*A*T' = [ A11  A12   0  ...   0   ;
                    A21  A22  A23 ...   0   ;
                    ...  ...  ... ...  ...  ;
                     *    *    *  ... Ak-1k ;
                    Ak1  Ak2  Ak3 ...  Akk  ]

    and

        T*B = [0; B0],

    where Aij is a block of size (ni x nj) and B0 is full row rank (nk x m).

    Parameters
    ----------
    A : (n, n) ndarray
        System matrix.
    B : (n, m) ndarray
        Control matrix.
    tol : float, optional
        Rank tolerance (default 1e-6).

    Returns
    -------
    T : (n, n) ndarray or 0
        Orthogonal transformation matrix. 0 if system is not controllable.
    nt : (k,) ndarray
        List of block sizes [n1, ..., nk].
    """
    n = A.shape[0]

    # --- Controllability check ---
    U = np.zeros((n, 0))
    Ak = np.eye(n)
    for i in range(n):
        U = np.hstack([U, Ak @ B])
        Ak = Ak @ A

    if np.linalg.matrix_rank(U, tol) < n:
        print("The pair {A,B} is not controllable.")
        return 0, np.array([])

    # --- Staircase orthogonal transformation ---
    T = np.eye(n)
    Ak_mat = A.copy()
    Bk = B.copy()
    l = 0
    nt_list = []

    while np.linalg.matrix_rank(Bk, tol) < Ak_mat.shape[0]:
        rk = np.linalg.matrix_rank(Bk, tol)
        nt_list.insert(0, rk)

        # Orthogonal complement of Bk (rows of B_ort span the nullspace of Bk')
        # null_space(Bk.T) gives orthonormal basis for nullspace of Bk'
        B_ort = null_space(Bk.T).T   # shape (nk - rk, nk)
        # Orthogonal complement of B_ort (rows span the rowspace of Bk')
        B_p = null_space(B_ort).T    # shape (rk, nk)

        if Ak_mat.shape[0] < n:
            # MATLAB: [[B_ort; B_p] zeros(nk,l); zeros(l,nk) eye(l)]
            # B_ort (nk-rk, nk) stacked vertically with B_p (rk, nk) → (nk, nk)
            top_left = np.vstack([B_ort, B_p])
            top_row = np.hstack([top_left, np.zeros((Ak_mat.shape[0], l))])
            bottom_row = np.hstack([np.zeros((l, Ak_mat.shape[0])), np.eye(l)])
            T = np.vstack([top_row, bottom_row]) @ T
        else:
            T = np.vstack([B_ort, B_p])

        l += rk
        Bk = B_ort @ Ak_mat @ B_p.T
        Ak_mat = B_ort @ Ak_mat @ B_ort.T

    # The final block is the last Bk
    nt_list.insert(0, np.linalg.matrix_rank(Bk, tol))
    nt = np.array(nt_list, dtype=int)

    return T, nt

File: HCS_Toolbox_ver02/trans_con.m
function [T nt]=trans_con(A,B)
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%  The function [T n_list]=trans_con(A,B) computes           
%%  an orthogonal matrix T for a controllable pair {A,B}:     
%%     T*A*T' = [ A11  A12   0  ...   0  ;                    
%%                A21  A22  A23 ...   0  ;                    
%%                ...  ...  ... ...  ... ;                    
%%                 *    *    *  ... Ak-1k;                    
%%                Ak1  Ak2  Ak3 ...  Akk ]                    
%%  and                                                       
%%              T*B=[0; B0],                                  
%%  where                                                     
%%    - Aij is a block of the size (ni x nj) with i,j=1,...,k   
%%    - B0 is a full row rank  matrix (nk x m)                
%%  The array n_list=[n1,...,nk] contains sizes of the blocks.  
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
n=size(A,1);nt=[]; T=0;
if n~=size(A,2) disp('Error: The matrix A is not square.'); return; end;
if n~=size(B,1) disp('Error: The number of rows of the matrix A does not coincide with the number of rows of the matrix B.'); return; end;

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%-----Controlability check----------
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
tol=0.000001;
U=[];Ak=eye(n);
for i=1:n
U=[U Ak*B];
Ak=Ak*A;
end;
if rank(U,tol)<n disp('The pair {A,B} is not controlable.'); return;
end;
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%----Transformation------
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
T=eye(n);
Ak=A;Bk=B;l=0;
while rank(Bk)<size(Ak,1)
    nt=[rank(Bk); nt];
    B_ort=null(Bk')';
    B_p=null(B_ort)';
    if size(Ak,1)<n T=[[B_ort;B_p] zeros(size(Ak,1),l); zeros(l,size(Ak,1)) eye(l)]*T;
    else T=[B_ort;B_p];
    end;
    l=l+rank(Bk);
    Bk=B_ort*Ak*B_p';
    Ak=B_ort*Ak*B_ort';        
end;
nt=[rank(Bk); nt];